vsintegration/src/FSharp.LanguageService/FSharpProjectSite.fs

```diff
@@ -0,0 +1,98 @@
+
+    member internal this.CheckOptions = 
+            match checkOptions with 
+            | Some options -> options
+            | None -> failwith "Options haven't been computed yet."
+
+    member internal this.Initialize(hier: IVsHierarchy, site : IProjectSite) =
+        this.ProjectTracker.AddProject(this)
+
+        site.AdviseProjectSiteChanges(FSharpCommonConstants.FSharpLanguageServiceCallbackName,
+                                        new AdviseProjectSiteChanges(fun () -> this.OnProjectSettingsChanged(hier, site)))
+
+        site.AdviseProjectSiteClosed(FSharpCommonConstants.FSharpLanguageServiceCallbackName, 
+                                        new AdviseProjectSiteChanges(fun () -> this.Disconnect()))
+
+        // Add files and references.
+        site.SourceFilesOnDisk() |> 
+            Seq.iter(fun file -> this.AddDocument(hier, file))
+        
+        this.GetReferences(site.CompilerFlags()) |> 
+            Seq.iter(fun ref -> this.AddReference(ref) |> ignore)
+        
+        // Capture the F# specific options that we'll pass to the type checker.
+        checkOptions <- Some(ProjectSitesAndFiles.GetProjectOptionsForProjectSite(site, site.ProjectFileName()))
+
+    member this.GetReferences(flags : string[]) =
+        let (|Reference|_|) (f : string) = if f.StartsWith("-r:") then Some (f.Replace("-r:", "")) else None
+
+        let references = flags |> 
+                         Seq.map(fun flag -> match flag with 
+                                             | Reference ref -> ref
+                                             | _ -> "") |>
+                         Seq.where(fun s -> s <> "")
+        references
+
+    member this.AddReference(filePath : string) = 
+        this.AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(filePath, new MetadataReferenceProperties(), VSConstants.S_FALSE)
+
+    member this.RemoveReference(filePath: string) =
+        this.RemoveMetadataReference(filePath)
+
+    member internal this.AddDocument(hier: IVsHierarchy, file : string) = 
+        let itemid = 
+            match hier.ParseCanonicalName(file) with
+            | (VSConstants.S_OK, id) -> id
+            | _ -> uint32 VSConstants.VSITEMID.Nil
+
+        let document = this.ProjectTracker.DocumentProvider.TryGetDocumentForFile(this, itemid, file, SourceCodeKind.Regular, fun x -> true)
+        this.AddDocument(document, true)
+
+    member internal this.OnProjectSettingsChanged(hier: IVsHierarchy, site : IProjectSite) = 
+        let sourceFiles = site.SourceFilesOnDisk()
+        // Added files.
+        sourceFiles |> 
+            Seq.where(fun file -> not(this.ContainsFile(file))) |> 
+            Seq.iter(fun file -> this.AddDocument(hier, file))
+        // Removed files.
+        this.GetCurrentDocuments() |> 
+            Seq.where(fun doc -> not(sourceFiles |> Seq.contains(doc.FilePath))) |>
+            Seq.iter(fun doc -> this.RemoveDocument(doc))
```

Comment:
Calling `RemoveDocument` while iterating over `GetCurrentDocuments` is fragile since it assumes that `GetCurrentDocuments` is returning a copy. Instead, consider looping over `GetCurrentDocuments` to create a list of files to remove, then looping over that list calling `RemoveDocument`.